rs/tests/nested/src/lib.rs

```diff
@@ -327,130 +453,251 @@ pub fn nns_recovery_test(env: TestEnv) {
+        f + 1
     );
-
-    // SSH into all guestOS nodes and break the replica
+    let faulty_nodes = nns_nodes.take(f + 1);
     let ssh_command =
         "sudo mount --bind /bin/false /opt/ic/bin/replica && sudo systemctl restart ic-replica";
-
-    for (i, node) in nns_nodes.iter().enumerate() {
+    for node in faulty_nodes {
         info!(
             logger,
             "Breaking the replica on node {} ({:?})...",
-            i + 1,
+            node.node_id,
             node.get_ip_addr()
         );
 
         node.block_on_bash_script(ssh_command).unwrap_or_else(|_| {
             panic!(
                 "SSH command failed on node {} ({:?})",
-                i + 1,
+                node.node_id,
                 node.get_ip_addr()
             )
         });
     }
 
-    // Test if the subnet is broken by trying to store a new message
+    info!(logger, "Ensure a healthy node still works in read mode");
+    assert!(can_read_msg(
+        &logger,
+        &dfinity_owned_node.get_public_url(),
+        app_can_id,
+        msg
+    ));
     info!(
         logger,
-        "Testing if the subnet is broken by attempting to store a new message..."
-    );
-    assert!(
-        cannot_store_msg(
-            logger.clone(),
-            &nns_node.get_public_url(),
-            test_can_id,
-            "subnet broken test message"
-        ),
-        "Subnet is still functional after breaking attempt - storing messages still succeeds",
+        "Ensure the subnet does not work in write mode anymore"
     );
+    assert!(cannot_store_msg(
+        logger.clone(),
+        &dfinity_owned_node.get_public_url(),
+        app_can_id,
+        msg
+    ));
     info!(
         logger,
         "Success: Subnet is broken - cannot store new messages"
     );
 
-    info!(logger, "Verifying that read operations still work...");
-    let can_read = can_read_msg(&logger, &nns_node.get_public_url(), test_can_id, test_msg);
+    info!(
+        logger,
+        "Starting recovery of the NNS subnet {}",
+        topology_after_removal.root_subnet_id().to_string()
+    );
 
-    if can_read {
-        info!(
-            logger,
-            "Success: Read operations still work as expected in broken subnet"
-        );
-    } else {
-        // QUESTION FOR PIERUGO: should this be a warning or an error? In the consensus recovery tests I looked at, we expect the read to succeed after breaking the subnet, but when I run this test, the read fails.
-        info!(
-            logger,
-            "WARNING: Read operations also failed - this might indicate a different issue"
-        );
+    // go over all steps of the NNS recovery
+    for (step_type, step) in subnet_recovery {
+        info!(logger, "Next step: {:?}", step_type);
+
+        info!(logger, "{}", step.descr());
+        step.exec()
+            .unwrap_or_else(|e| panic!("Execution of step {:?} failed: {}", step_type, e));
     }
+    info!(logger, "NNS recovery has finished");
```

Comment:
Maybe prep isn't the right term, but I found this confusing. When I read recovery has finished, I think that the recovery has finished.

```suggestion
    for (step_type, step) in subnet_recovery_prep {
        info!(logger, "Next step: {:?}", step_type);

        info!(logger, "{}", step.descr());
        step.exec()
            .unwrap_or_else(|e| panic!("Execution of step {:?} failed: {}", step_type, e));
    }
    info!(logger, "NNS recovery prep has finished");
```